### PR TITLE
[ new ] Closure of a Pred wrt a (pre/strict partial)order

### PR DESCRIPTION
**Pull request by gallais**
I started adding Box & Thinnable (aka `Closed` for the preorder `_\sub=_`) to `Data.List.Relation.Sublist.Inductive` and then realized we could generalize
the whole setup.

### CODE REVIEW

File: src/Relation/Unary/Closure.agda [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+open import Relation.Binary as Rel hiding (module Preorder; module StrictPartialOrder)
+
+module Preorder {a r e} (P : Preorder a e r) where
+  open Rel.Preorder P
+
+  -- Depending on whether the order is increasing or decreasing, we get either
+  -- the definition of the upward-closed subsets or that of the downward-closed
+  -- ones.
+  □ : ∀ {t} → Pred Carrier t → Pred Carrier (a ⊔ r ⊔ t)
+  □ T x = ∀ {y} → x ∼ y → T y
```
Comment (MatthewDaggitt) [inline src/Relation/Unary/Closure.agda]: I'm missing the character set for what you've called this type. Any idea what I need to install?

File: src/Relation/Unary/Closure.agda [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  -- Closed: whenever we have a value in one context, we can get one in any
+  -- related context.
+  record Closed {t} (T : Pred Carrier t) : Set (a ⊔ r ⊔ t) where
+    field next : ∀ {x} → T x → □ T x
+
+  -- □ is the Closure operator i.e. for any `T`, `□ T` is closed.
+  □-closed : ∀ {t} {T : Pred Carrier t} → Closed (□ T)
+  □-closed = record { next = duplicate }
+
+module StrictPartialOrder {a r e} (P : StrictPartialOrder a e r) where
```
Comment (MatthewDaggitt) [inline src/Relation/Unary/Closure.agda]: Should these go in their own top level module. `Closure.Preorder` and `Closure.StrictPartialOrder`?

File: src/Relation/Unary/Closure.agda (hunk 1 above)
Comment (gallais) [reply src/Relation/Unary/Closure.agda]: It's this character: https://www.fileformat.info/info/unicode/char/25a1/index.htm

We picked it in https://dl.acm.org/citation.cfm?id=3236785 by analogy to modal
logic where box also means "for all successors".

File: src/Relation/Unary/Closure.agda (hunk 2 above)
Comment (gallais) [reply src/Relation/Unary/Closure.agda]: I wasn't too sure whether I should split them off or not. I'll do so.

File: src/Relation/Unary/Closure.agda (hunk 1 above)
Comment (JacquesCarette) [reply src/Relation/Unary/Closure.agda]: It's a box... which might show up looking quite similar [might even be identical] to the box you get if you don't have the font installed. I am not sure how I would debug this!

**Comment by gallais** [comment]
We have two kind of maps:

1. `x ~ y -> □ T x -> □ T y`
2. `(T => U) -> (□ T => □ U)`

Not sure which names to use.

Maybe `reindex` for 1. and `map` for 2.?

File: src/Relation/Unary/Closure.agda (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Relation/Unary/Closure.agda]: That's exactly the symbol that's used to render characters that aren't installed on my system :laughing: 

Can you think of any other alternatives names? Might be a little confusing to people...

Comment (gallais) [reply src/Relation/Unary/Closure.agda]: The connection to modal logic really calls for box though:

> w ⊨ □ P if and only if for every element u of G, if w R u then u ⊨ P ([source](https://en.wikipedia.org/wiki/Modal_logic))

vs.

> □ T x = ∀ {y} → x ∼ y → T y

**Comment by MatthewDaggitt** [comment]
> Maybe `reindex` for 1. and `map` for 2.?

Sounds good.

File: src/Relation/Unary/Closure.agda (hunk 1 above)
Comment (JacquesCarette) [reply src/Relation/Unary/Closure.agda]: I think @MatthewDaggitt means can you pick a different box?

**Comment by JacquesCarette** [comment]
I agree as well.

File: src/Relation/Unary/Closure.agda (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Relation/Unary/Closure.agda]: I guess. Or we need to make it very clear at the top of the file that the user *isn't* missing a load of fonts!

Comment (gallais) [reply src/Relation/Unary/Closure.agda]: Oh, I see. Looking at the input method for the ones available, which one do you prefer?

1. □ -- WHITE SQUARE (the current one)
2. ◻ -- WHITE MEDIUM SQUARE
3. ◽ -- WHITE MEDIUM SMALL SQUARE

Wikipedia seems to advocate either 1. or 2. https://en.wikipedia.org/wiki/List_of_logic_symbols

Comment (MatthewDaggitt) [reply src/Relation/Unary/Closure.agda]: 1. and 2. are being rendered identically for me. 3. is noticeably smaller but I guess it's not recommended. 

Hmm I'm not quite sure what to do....

Comment (gallais) [reply src/Relation/Unary/Closure.agda]: > Hmm I'm not quite sure what to do....

I would go with your previous suggestion of adding a preamble at the top
of the file.

It's also a good opportunity to throw in a paragraph or two to give the
reader an intuition of what these mean and clarify the connection with
modal logic, kripke models, etc.

**Comment by JacquesCarette** [comment]
Aren't these two new modules essentially identical? This seems like a lot of code duplication.

**Pull request merged**